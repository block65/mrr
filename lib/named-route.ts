import type { Params, ExtractRouteParams } from './types.js';
import { nullOrigin, urlObjectAssign, urlRhs } from './util.js';

type QueryParams = Record<string, string>;

export interface NamedRoute<P extends string, Q extends QueryParams> {
  path: P;
  build: (x?: {
    params?: ExtractRouteParams<P>;
    query?: Q;
    hash?: string;
    origin?: string;
  }) => string;
}

export function interpolate<P extends string>(
  path: P,
  params: ExtractRouteParams<P> | Params = {},
): string {
  return path.replace(/\/:(\w+)[?+*]?/g, (_match, token: keyof typeof params) =>
    params[token] ? `/${params[token]}` : '',
  );
}

export function namedRoute<T extends string, Q extends QueryParams = never>(
  pathname: T,
): NamedRoute<T, Q> {
  return {
    path: pathname,
    build({ params, query, origin, ...rest } = {}) {
      const interpolatedPathname = params
        ? interpolate(pathname, params)
        : pathname;

      const searchParams = new URLSearchParams(query);
      searchParams.sort();

      const newUrl = urlObjectAssign(new URL(origin || nullOrigin), {
        pathname: interpolatedPathname,
        search: searchParams.toString(),
        ...rest,
      });

      return newUrl.origin === nullOrigin.origin
        ? urlRhs(newUrl)
        : newUrl.toString();
    },
  };
}
