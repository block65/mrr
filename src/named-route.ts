export { namedRoute } from '../lib/named-route.js';

export type {
  NamedRoute,
  NamedRouteWithParams,
  NamedRouteWithoutParams,
} from '../lib/named-route.js';
