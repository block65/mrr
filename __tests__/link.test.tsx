import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import type { AnchorHTMLAttributes, FC, PropsWithChildren } from 'react';
import { FormattedMessage, IntlProvider } from 'react-intl';
import { namedRoute } from '../lib/named-route.js';
import { Link, Router, useLocation } from '../src/index.js';

export const LocationDisplay = () => {
  const [location] = useLocation();

  return <div data-testid="location-display">{location.pathname}</div>;
};

const LinkyLink: FC<
  PropsWithChildren<AnchorHTMLAttributes<HTMLAnchorElement>>
> = (props) => <span data-linkylink {...props} />;

const CompyComp: FC<PropsWithChildren> = ({ children, ...props }) => (
  <div data-compycomp {...props}>
    {children}
  </div>
);

const BrokeyBroke: FC<PropsWithChildren> = ({ children }) => <>{children}</>;

test('basic', async () => {
  const login = namedRoute('/login');

  const { asFragment } = render(
    <IntlProvider locale="en" onError={() => {}}>
      <Router origin="https://www.example.com">
        <Link dest={login.build()}>test as string</Link>
        <Link dest="https://invalid.example.com">cross origin link</Link>
        <Link dest={login.build()}>
          <>test as fragment</>
        </Link>
        <Link dest={login.build()}>
          <LinkyLink>test nested in an FC</LinkyLink>
        </Link>
        <Link dest={login.build()}>
          <FormattedMessage
            id="broken"
            description="this will not work properly"
            defaultMessage="this will not work properly"
          />
        </Link>

        <Link dest={login.build()}>plnk plonk</Link>

        <Link dest={login.build()}>
          <a id="123">Hello!</a>
        </Link>

        <Link dest={login.build()}>
          <CompyComp>
            Once
            <CompyComp>
              Twice
              <CompyComp>Three times a lady</CompyComp>
            </CompyComp>
          </CompyComp>
        </Link>
        <Link dest={login.build()}>
          <BrokeyBroke>this will also not work properly</BrokeyBroke>
        </Link>
      </Router>
      ,
    </IntlProvider>,
  );

  expect(asFragment()).toMatchSnapshot();
});
