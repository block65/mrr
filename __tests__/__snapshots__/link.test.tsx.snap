// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic 1`] = `
<DocumentFragment>
  <a
    href="/login"
  >
    test as string
  </a>
  <a
    href="/login"
  >
    test as fragment
  </a>
  <a
    data-linkylink="true"
    href="/login"
  >
    this text should be in a component that has a href
  </a>
  this text will not be in a component that has a href :sadface:
  <a
    href="/login"
    id="123"
  >
    This anchor should have a href!
  </a>
  this text will not be in a component that has a href :sadface:
  <a
    href="https://invalid.example.com"
    rel="no-opener noreferrer"
  >
    cross origin link
  </a>
</DocumentFragment>
`;
